### PR TITLE
UCP/GTEST: Force creating an am_lane for the client-server flow.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6666/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6667/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@brminich can you please review?

**Comment by brminich** [comment]
sure, its WIP ;)

File: src/ucp/core/ucp_proxy_ep.c [hunk 1]
```diff
@@ -183,20 +183,44 @@ static UCS_CLASS_CLEANUP_FUNC(ucp_proxy_ep_t)
...
     ucp_ep_h ucp_ep = proxy_ep->ucp_ep;
     ucp_lane_index_t lane;
+    uct_ep_h tl_ep = NULL;
 
     ucs_assert(proxy_ep->uct_ep != NULL);
     for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
         if (ucp_ep->uct_eps[lane] == &proxy_ep->super) {
+            ucs_assert(proxy_ep->uct_ep != NULL);    /* make sure there is only one match */
             ucp_ep->uct_eps[lane] = proxy_ep->uct_ep;
+            tl_ep = ucp_ep->uct_eps[lane];
```
Comment (brminich) [inline src/ucp/core/ucp_proxy_ep.c]: why do you need to go further? (break is removed)

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 2]
```diff
@@ -246,4 +270,41 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    struct sockaddr_in connect_addr;
+    get_listen_addr(&connect_addr);
+
+    UCS_TEST_MESSAGE << "Testing " << ucs::sockaddr_to_str((const struct sockaddr*)&connect_addr);
+
+    start_listener((const struct sockaddr*)&connect_addr);
+
+    client_ep_connect((struct sockaddr*)&connect_addr);
+
+    while (receiver().get_num_eps() == 0) {
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: maybe add timeout here to avoid hang in case something goes wrong

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -734,23 +759,13 @@ static ucs_status_t ucp_wireup_add_am_lane(ucp_ep_h ep, const ucp_ep_params_t *p
...
-        (ep_init_flags & UCP_EP_INIT_FLAG_MEM_TYPE)) {
-        need_am = 0;
-        for (lane = 0; lane < *num_lanes_p; ++lane) {
-            need_am = need_am || ucp_worker_is_tl_p2p(ep->worker,
-                                                      lane_descs[lane].rsc_index);
-        }
-        if (!need_am) {
-            return UCS_OK;
-        }
+    if (!ucp_wireup_is_am_config_required(ep, params, ep_init_flags, lane_descs,
```
Comment (brminich) [inline src/ucp/wireup/select.c]: minor: maybe something like `am_lane_required`? 
`config` looks a bit confusing

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 4]
```diff
@@ -246,4 +270,41 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+        check_events(sender().worker(), receiver().worker(), false);
+    }
+
+    /* allow the connection establishment flow to complete */
+    short_progress_loop();
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: why is this needed?

File: test/gtest/ucp/ucp_test.h [hunk 5]
```diff
@@ -221,5 +220,8 @@ std::ostream& operator<<(std::ostream& os, const ucp_test_param& test_param);
     UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, self,   "self") \
     UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, tcp,    "tcp")
 
+#define UCP_INSTANTIATE_ALL_TEST_CASE(_test_case) \
```
Comment (brminich) [inline test/gtest/ucp/ucp_test.h]: maybe add this to test_ucp_sockaddr.cc instead?

File: src/ucp/core/ucp_proxy_ep.c (hunk 1 above)
Comment (brminich) [reply src/ucp/core/ucp_proxy_ep.c]: I just thought it should never happen. Ok, let's have an assert

File: test/gtest/ucp/ucp_test.h (hunk 5 above)
Comment (brminich) [reply test/gtest/ucp/ucp_test.h]: then maybe we need to run all tests with new TL?

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 4 above)
Comment (brminich) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: How do you ensure that the connection is ready?

File: src/ucp/core/ucp_proxy_ep.c (hunk 1 above)
Comment (alinask) [reply src/ucp/core/ucp_proxy_ep.c]: right, the assert will make sure of that.

File: test/gtest/ucp/ucp_test.h (hunk 5 above)
Comment (alinask) [reply test/gtest/ucp/ucp_test.h]: not sure becuase the other tests run shared memory with ib (shm_ib) which works for the other tests but doesn't really cover the mm case for the client-server flow due to a large worker address - which will be addressed in the future. so in this case i need to make the address smaller so that the test would work with rma and mm - with xpmem.
```
alinas@hpc-test-node ~/workspace/tmp/ucx
$UCX_SHM_DEVICES=xpmem ./test/gtest/gtest --gtest_filter=mm_rdmacm/test_ucp_sockaddr_with_rma_atomic.wireup_for_rma_atomic/0
[     INFO ] checking dc: yes
[     INFO ] checking dcx: yes
[     INFO ] checking ud: yes
[     INFO ] checking udx: yes
[     INFO ] checking rc: yes
[     INFO ] checking rcx: yes
[     INFO ] checking rcx_cm: yes
[     INFO ] checking shm_ib: yes
[     INFO ] checking ugni: no
[     INFO ] checking self: yes
[     INFO ] checking tcp: yes
[     INFO ] checking all_rcdc: yes
[     INFO ] checking all: yes
[     INFO ] checking rc_x: yes
[     INFO ] checking ud_mlx5: yes
[     INFO ] checking shm: yes
[     INFO ] checking mm_rdmacm: yes
[     INFO ] Using random seed of 3258
Note: Google Test filter = mm_rdmacm/test_ucp_sockaddr_with_rma_atomic.wireup_for_rma_atomic/0
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from mm_rdmacm/test_ucp_sockaddr_with_rma_atomic
[ RUN      ] mm_rdmacm/test_ucp_sockaddr_with_rma_atomic.wireup_for_rma_atomic/0
[     INFO ] Testing 1.1.1.5:47408
[       OK ] mm_rdmacm/test_ucp_sockaddr_with_rma_atomic.wireup_for_rma_atomic/0 (106 ms)
[----------] 1 test from mm_rdmacm/test_ucp_sockaddr_with_rma_atomic (106 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (107 ms total)
[  PASSED  ] 1 test.
``` 
other tests don't have this limitation so shared memory is tested in them anyway.

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 4 above)
Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: on the client side we don't have a callback or anything to make sure of that. but this test checks that there is an am_lane for wireup. so for the server side if i remove the forcing of an am_lane, the test will fail on not having an am_lane for wireup - as expected. for client - for connect_to_iface transports, the dont really need an am_lane (for now) since they can connect to the server right away (after getting a wireup_conn_request from it). for connect_to_ep transports, they will have an am_lane regardless of the change in this pr so this test is not intended for testing this.
buttom line, i can remove the forcing of an am_lane for the client and it would still work but we decided to do this anyway since it will probably be needed in the future.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6680/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe can you please review?

File: src/ucp/core/ucp_proxy_ep.c [hunk 6]
```diff
@@ -183,20 +183,44 @@ static UCS_CLASS_CLEANUP_FUNC(ucp_proxy_ep_t)
     }
 }
 
+static void ucp_proxy_ep_replace_if_owned(uct_ep_h uct_ep, uct_ep_h owned_ep,
+                                          uct_ep_h replacement_ep)
+{
+    ucp_proxy_ep_t *proxy_ep;
+
+    if (uct_ep->iface->ops.ep_destroy == ucp_proxy_ep_destroy) {
```
Comment (yosefe) [inline src/ucp/core/ucp_proxy_ep.c]: pls add a function ucp_ep_proxy_test() to wrap this check

File: src/ucp/core/ucp_proxy_ep.c [hunk 7]
```diff
@@ -183,20 +183,44 @@ static UCS_CLASS_CLEANUP_FUNC(ucp_proxy_ep_t)
...
+            ucs_assert(proxy_ep->uct_ep != NULL);    /* make sure there is only one match */
             ucp_ep->uct_eps[lane] = proxy_ep->uct_ep;
+            tl_ep = ucp_ep->uct_eps[lane];
             proxy_ep->uct_ep = NULL;
-            break;
         }
     }
 
+    /* go through the lanes and check if the wireup_ep which will be destroyed,
+     * is pointed to by another proxy ep. if so, set the wireup ep's real tl
```
Comment (yosefe) [inline src/ucp/core/ucp_proxy_ep.c]: this is generic code for proxy ep, not just for wireup
"go through the lane and check if the proxy ep that is being destroyed is pointer to by another proxy wp. if so, redirect that other proxy ep to point to the underlying uct ep"

File: src/ucp/wireup/select.c [hunk 8]
```diff
@@ -724,6 +724,31 @@ static int ucp_wireup_is_lane_proxy(ucp_ep_h ep, ucp_rsc_index_t rsc_index,
...
+                                            unsigned ep_init_flags,
+                                            ucp_wireup_lane_desc_t *lane_descs,
+                                            int num_lanes_p)
+{
+    ucp_lane_index_t lane;
+    int need_am = 1;
+
+    /* Check if we need active messages from the configurations, for wireup.
+     * If not, check if am is required due to p2p transports */
+    if ((!(ucp_ep_get_context_features(ep) & (UCP_FEATURE_TAG | UCP_FEATURE_STREAM)) ||
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: could be written in better way:
```
if (flags & AM_LANE) {
    return 1;
}

if (!(flags & MEM_TYPE) && (featrures & (TAG|STREAM)) {
    return 1;
}

for (lane = 0; ...) {
  if (worker_is_tl_p2p()) {
    return 1;
  }
}
return 0;
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6687/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6691/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe please review again.
